server/spec/services/scheduler/strategy/random_spec.rb

```diff
@@ -8,21 +7,60 @@
     nodes << HostNode.create!(node_id: 'node1', name: 'node-1', connected: true, grid: grid)
     nodes << HostNode.create!(node_id: 'node2', name: 'node-2', connected: true, grid: grid)
     nodes << HostNode.create!(node_id: 'node3', name: 'node-3', connected: true, grid: grid)
-    nodes
+    nodes.map { |n| Scheduler::Node.new(n) }
   end
 
   let(:stateful_service) do
-    GridService.create!(name: 'test', grid: grid, image_name: 'foo/bar:latest', stateful: true)
+    GridService.create!(
+      strategy: 'random', name: 'test', grid: grid, image_name: 'foo/bar:latest', stateful: true
+    )
   end
 
   let(:stateless_service) do
-    GridService.create!(name: 'test', grid: grid, image_name: 'foo/bar:latest', stateful: false)
+    GridService.create!(
+      strategy: 'random', name: 'test', grid: grid, image_name: 'foo/bar:latest', stateful: false
+    )
   end
 
   describe '#find_node' do
-    it 'returns random node' do
-      node = subject.find_node(stateless_service, 2, nodes)
-      expect(nodes.include?(node)).to be_truthy
+    context 'stateless' do 
+      it 'returns a node if no previous instance exist' do
+        node = subject.find_node(stateless_service, 2, nodes)
+        expect(nodes.include?(node)).to be_truthy
+      end
+
+      it 'returns previously scheduled node if it exists' do
+        stateless_service.grid_service_instances.create!(instance_number: 2, host_node: nodes[2].node)
+        node = subject.find_node(stateless_service, 2, nodes)
+        expect(node).to eq(nodes[2])
+      end
+
+      it 'returns a node if previously scheduled node is not included in nodes array' do
+        stateless_service.grid_service_instances.create!(instance_number: 2, host_node: nodes[2].node)
+        nodes.delete(nodes[2])
+        node = subject.find_node(stateless_service, 2, nodes)
+        expect(node).not_to eq(nodes[2])
+      end
+    end
+
+    context 'stateful' do
+      it 'returns a node if no previous instance exist' do
+        node = subject.find_node(stateful_service, 2, nodes)
+        expect(nodes.include?(node)).to be_truthy
+      end
+
+      it 'returns previously scheduled node if it exists' do
+        stateful_service.grid_service_instances.create!(instance_number: 2, host_node: nodes[2].node)
+        node = subject.find_node(stateful_service, 2, nodes)
+        expect(node).to eq(nodes[2])
+      end
+
+      it 'returns nil if previously scheduled node is not included in nodes array' do
```

Comment:
This is different from how the `ha` strategy behaves, though? If the host node that the stateful instance is deployed to is removed, or filtered away, then it just returns a new node to reschedule the stateful instance onto?

This would fail the deploy with an error in that case?